src/main/scala/io/iohk/ethereum/network/discovery/PeerDiscoveryManager.scala

```diff
@@ -145,34 +180,73 @@ class PeerDiscoveryManager(
   def sendDiscoveredNodesInfo(
       maybeDiscoveryService: Option[v4.DiscoveryService],
       recipient: ActorRef
-  ): Unit = {
+  ): Unit = pipeToRecipient(recipient) {
 
     val maybeDiscoveredNodes: Task[Set[Node]] =
       maybeDiscoveryService.fold(Task.pure(Set.empty[Node])) {
         _.getNodes.map { nodes =>
-          nodes.map { node =>
-            Node(
-              id = ByteString(node.id.toByteArray),
-              addr = node.address.ip,
-              tcpPort = node.address.tcpPort,
-              udpPort = node.address.udpPort
-            )
-          }
+          nodes.map(toNode)
         }
       }
 
     maybeDiscoveredNodes
       .map(_ ++ alreadyDiscoveredNodes)
-      .map(_.filterNot(_.id == localNodeId))
+      .map(_.filterNot(isLocalNode))
       .map(DiscoveredNodesInfo(_))
+  }
+
+  def sendRandomNodeInfo(
+      maybeRandomNodes: Option[RandomNodes],
+      recipient: ActorRef
+  ): Unit = pipeToRecipient[RandomNodeInfo](recipient) {
+    val randomNode = maybeRandomNodes match {
+      case None if alreadyDiscoveredNodes.isEmpty =>
+        Task.raiseError(
+          new IllegalStateException("There are no known nodes and discovery isn't running.")
+        )
+
+      case None =>
+        Task.pure(alreadyDiscoveredNodes(Random.nextInt(alreadyDiscoveredNodes.size)))
+
+      case Some(consumer) =>
+        consumer.pull
+          .flatMap {
+            case Left(None) =>
+              Task.raiseError(new IllegalStateException("The random node source is finished."))
```

Comment:
Maybe I'm missing something, but don't this exceptions result in Status.Failure being sent to the PeerManagerActor when piping the future to them?

If so, wouldn't it be better to handle them here and not send anything if that happened?

Reply:
Correct, if an exception is raised, `Status.Failure` is sent to the calling actor, which in our case is ignored by the `PeerManagerActor`. The error would be logged by `pipeToRecipient`, but this situation will never happen as the `Iterant` will never finish, it will keep repeating itself. I only included it here to handle all cases for the compiler, there's no need to handle anything, as it would be a programming error if this happened. I found thought that returning errors helps in tests though, that instead of a timeout we get to see the error message when using `ask` patterns.